Rename candidate center id to registration center

cmadjar opened the pull request:
### Brief summary of changes

This PR rename the `CenterID` field column of the `candidate` table to `RegistrationCenterID`, which will correspond to the `CenterID` at which the candidate was first registered. This should never be changed for a given candidate. This is part 1 of the refactoring of the candidate's CenterID. (See #4058 for details on part 2 of the refactoring).

### Associate LORIS-MRI PR

https://github.com/aces/Loris-MRI/pull/341 (@nicolasbrossard will test it)

### This resolves issue...

- [x] Github? #4057

### To test this change...

- [ ] Make sure the following pages load and work as expected: 
> - [ ] dashboard 
> - [x] new profile @ridz1208
> - [x] candidate list @ridz1208
> - [x] timepoint list and create timepoint @ridz1208
> - [x] conflict resolver @ridz1208
> - [x] genomic_browser @xlecours 
> - [x] imaging browser  @davidblader 
> - [x] issue tracker @zaliqarosli 
> - [x] media @zaliqarosli 
> - [x] MRI violations @nicolasbrossard  
> - [x] statistics @ridz1208 
> - [x] survey accounts @sruthymathew123 
> - [x] API v0.0.2 and v.0.0.3-dev @driusan @xlecours 
> - [x] RasinBread dataset and test units @kongtiaowang 
> - [ ] tools
- [x] Test the SQL patch @PapillonMcGill 

**Notes for reviewers**

I assigned above several people to test various parts of LORIS that were affected by the changes based on my knowledge of who worked on which module. Figured it would be less overwhelming if several people tested one part of the PR instead of having to test everything. 

Note that since I don't have any genomics data or media and issue tracker set up, I could not test **genomic browser**, **media** or **issue tracker**.

@driusan @ridz1208 I was not sure who to assign the dashboard, new profile, candidate list, timepoint list, conflict resolvers and the tools. Maybe you can help filling the gap? 
 
### Notes for existing projects

- `RegistrationCenterID` should never be changed for a given candidate. The `session` table will have the information about the latest `CenterID` the candidate has been seen.
- Make sure that `c.CenterID` is replaced by `c.RegistrationCenterID` in your project code queries.

zaliqarosli commented:
@cmadjar i'm trying to apply the patch and get the error `Can't write; duplicate key in table`. I can't spot anything wrong with the syntax but i'm assuming the FK isn't getting dropped properly?

cmadjar commented:
@zaliqarosli I modified the patch. I ran with the same issue but could never figure out why it failed when everything is done with one query. I just split the query in 3 statements and it worked for me. Give it a try. Thanks!!

PapillonMcGill commented:
@cmadjar @zaliqarosli When applying all 3 changes in a single command, the DB server commit only at the semi-column. Thus, when trying to change the column name, the FK is not yet drop, nor the when trying to add it back with the new column name.

Using 3 separate command (with auto-commit), ensure that the FK is first drop before renaming the column and the column is renamed before adding back the FK.

Check the code, look good with of course a few minor thing Travis pointed.

cmadjar commented:
@PapillonMcGill Thank you for the clarification. That makes a lot of sense!

zaliqarosli commented:
failing when testing Issue Tracker. Errors from ajax/EditIssue.php: `Data truncated for column 'centerID' at row 1` (lines 36 and 106), `Unknown column 'CenterID' in 'field list' (lines 206, 90, 36). I get these errors when creating/editing an issue

zaliqarosli commented:
@cmadjar could you rebase so changes to media module is included? there's a bug with hidden headers on this branch that was fixed on major

cmadjar commented:
@zaliqarosli rebased to the latest version of aces/major. Please review again when you can. Thanks!

zaliqarosli commented on modules/dashboard/php/dashboard.class.inc:
```diff
@@ -144,7 +144,7 @@ class Dashboard extends \NDB_Form
 		             LEFT JOIN candidate c ON (c.CandID=s.CandID)
 		             LEFT JOIN Project p ON (c.ProjectID=p.ProjectID ) 
-                     WHERE FIND_IN_SET(c.CenterID, :siteID) 
+                     WHERE FIND_IN_SET(s.CenterID, :siteID) 
```
i think using find_in_set in fine here. but even if not, this changes shouldnt be done in this PR which is only to rename center id to registration center id

PapillonMcGill replied:
will include it in https://github.com/aces/Loris/pull/4085 to fix them all in one place

cmadjar replied:
done

cmadjar commented on php/libraries/NDB_BVL_Feedback.class.inc:
```diff
@@ -469,7 +469,7 @@ class NDB_BVL_Feedback
         $query .= " WHERE ft.Active ='Y'";
         if (!$hasReadPermission===true) {
-            $query            .= " AND FIND_IN_SET (c.CenterID,:CentID)";
+            $query            .= " AND FIND_IN_SET (s.CenterID,:CentID)";
```
done

cmadjar commented on php/libraries/NDB_BVL_Feedback.class.inc:
```diff
@@ -582,7 +582,7 @@ class NDB_BVL_Feedback
         } else {
             $query            .= " AND ft.Active ='Y' 
-                                   AND FIND_IN_SET (c.CenterID,:CentID)";
+                                   AND FIND_IN_SET (s.CenterID,:CentID)";
```
done

zaliqarosli commented on modules/genomic_browser/php/cnv_browser.class.inc:
```diff
@@ -140,15 +140,16 @@ class CNV_Browser extends \NDB_Menu_Filter
         if (!$user->hasPermission('genomic_browser_view_allsites')) {
             $site_arr     = implode(",", $user->getCenterIDs());
-            $this->query .= " AND candidate.CenterID IN (" . $site_arr . ")";
+            $this->query .= " AND candidate.RegistrationCenterID IN ("
```
i think this should be the session centerID?

zaliqarosli replied:
or maybe not? im not too sure now, i don't know how genomic browser works

cmadjar replied:
I don't think there is any sessionID to attach genomic data. @xlecours could you confirm?

zaliqarosli commented on modules/genomic_browser/php/snp_browser.class.inc:
```diff
@@ -148,15 +148,16 @@ class SNP_Browser extends \NDB_Menu_Filter
             // restrict data to own site
             $site_arr     = implode(",", $user->getCenterIDs());
-            $this->query .= " AND candidate.CenterID IN (" . $site_arr . ")";
+            $this->query .= " AND candidate.RegistrationCenterID IN ("
```
should also maybe be the session centerID

cmadjar replied:
Same comment. I don't think there is any sessionID to attach genomic data. @xlecours could you confirm?

zaliqarosli commented on php/libraries/NDB_BVL_Feedback.class.inc:
```diff
@@ -469,7 +469,7 @@ class NDB_BVL_Feedback
         $query .= " WHERE ft.Active ='Y'";
         if (!$hasReadPermission===true) {
-            $query            .= " AND FIND_IN_SET (c.CenterID,:CentID)";
+            $query            .= " AND FIND_IN_SET (s.CenterID,:CentID)";
```
no space between SET and opening parenthesis. space between comma and colon

cmadjar replied:
done

zaliqarosli commented on php/libraries/NDB_BVL_Feedback.class.inc:
```diff
@@ -582,7 +582,7 @@ class NDB_BVL_Feedback
         } else {
             $query            .= " AND ft.Active ='Y' 
-                                   AND FIND_IN_SET (c.CenterID,:CentID)";
+                                   AND FIND_IN_SET (s.CenterID,:CentID)";
```
same comment as above

cmadjar replied:
done

cmadjar commented on modules/candidate_list/php/candidate_list.class.inc:
```diff
@@ -88,7 +88,7 @@ class Candidate_List extends \NDB_Menu_Filter
 
         $this->formToFilter = array(
-                               'centerID'     => 'c.CenterID',
+                               'centerID'     => 's.CenterID',
```
Right. Should be changed now.

cmadjar commented on modules/candidate_list/php/candidate_list.class.inc:
```diff
@@ -97,7 +97,7 @@ class Candidate_List extends \NDB_Menu_Filter
         $this->validFilters = array(
                                'pso.ID',
-                               'c.CenterID',
+                               's.CenterID',
```
done as well

cmadjar commented on modules/candidate_list/php/candidate_list.class.inc:
```diff
@@ -168,7 +168,7 @@ class Candidate_List extends \NDB_Menu_Filter
         $this->EqualityFilters    = array(
                                      's.SubprojectID',
-                                     'c.CenterID',
+                                     's.CenterID',
```
done

cmadjar commented:
@zaliqarosli Sorry, I dismissed your review by merging @xlecours PR to my branch (I missed one instance of CenterID in the genomic module)

sruthymathew123 commented:
@cmadjar Seems that this PR is affected in survey accounts with its automated integration test script.
It all looks good to me.

davidblader commented:
imaging browser is good

cmadjar commented:
> @cmadjar Seems that this PR is affected in survey accounts with its automated integration test script.
It all looks good to me.

@sruthymathew123 was it the survey accounts automated test script or the imaging browser? Looking at Travis' log, I spotted one error on the imaging browser test unit but did not see any problem with the survey. Maybe I am not looking at the right place? Thanks!

ridz1208 commented on modules/create_timepoint/php/create_timepoint.class.inc:
```diff
@@ -51,7 +51,11 @@ class Create_Timepoint extends \NDB_Form
             $user->hasPermission('data_entry') &&
-            (in_array($candidate->getData('CenterID'), $user->getData('CenterIDs')))
+            (in_array(
+                $candidate->getData('RegistrationCenterID'),
```
this is fine for now but should be bettered in phase 2

ridz1208 commented on modules/timepoint_list/php/timepoint_list.class.inc:
```diff
@@ -44,14 +44,14 @@ class Timepoint_List extends \NDB_Menu
         if ($user->hasPermission('access_all_profiles')
             || (in_array(
-                $candidate->getData('CenterID'),
+                $candidate->getData('RegistrationCenterID'),
```
not sure I agree with this one, I think we need to make it more intelligent and have it look for at lease one timepoint at the user's site.

also acceptable in phase 2

cmadjar replied:
copied a link to that comment in issue #4058 so that it gets tackled properly when coding phase 2. Thanks @ridz1208

ridz1208 commented on modules/timepoint_list/test/TestPlan.md:
```diff
@@ -2,7 +2,10 @@
-    - For a candidate of a different site than your user, ensure that `access_all_profiles` permission is required.
+    - For a candidate of a different site than your user, ensure that either 
+        - `access_all_profiles` permission is required 
+        - or that at least one of the candidate's sessions was performed at the 
```
I dont think that thats what the current code is doing. it seems like it's only if the candidate's registration site is in the list of sites of the user

cmadjar replied:
Right. Modifying the documentation accordingly.

ridz1208 commented:
@cmadjar 

new profile (there does not seem to be changes directly on the module)
candidate list (comment added)
timepoint list (comment added, phase 2 work involved)
create timepoint (phase 2 work involved, looks good for now)
conflict resolver (looks good)
statistics (comments added)

cmadjar commented:
@ridz1208 see answers to this comment below: 

> new profile (there does not seem to be changes directly on the module)
from what I understand of the module, it is by default using the CenterID of the user that is accessing this page. When the user belongs to multiple sites, it uses the first CenterID at which the user was created. So I don't think there is anything to change here.

> candidate list (comment added)
fixed the code in the last few commits

> timepoint list (comment added, phase 2 work involved)
fixed the code and also added link of your comment to issue #4058 so that it is implemented in phase 2

> create timepoint (phase 2 work involved, looks good for now)
added link of your comment to issue #4058 so that it is implemented in phase 2

> conflict resolver (looks good)
YEEEE!

> statistics (comments added)
should have fixed your comments now

Let me know if there is anything I missed or misunderstood. Thank you so much for having taken the time to review my PR in depth :)

ridz1208 commented on modules/statistics/php/stats_demographic.class.inc:
```diff
@@ -436,16 +436,16 @@ class Stats_Demographic extends \NDB_Form
                 AND f.CommentID NOT LIKE 'DDE%'
                 $paramProject
-                GROUP BY c.CenterID, SubprojectID, VLabel, Subcat
+                GROUP BY s.CenterID, SubprojectID, VLabel, Subcat
```
@cmadjar I dont think people would have fully understood the bullet points
we could bring them up with specific people if need be. but these changes are enough to pass review on this PR. I just wanted to confirm in the loris meeting that we should only be counting candidates once

ridz1208 commented on modules/statistics/php/stats_demographic.class.inc:
```diff
@@ -209,7 +209,7 @@ class Stats_Demographic extends \NDB_Form
 
         if (!empty($demographicSite)) {
-            $paramSite           = 'AND (c.CenterID IS NULL OR c.CenterID=:sid) ';
+            $paramSite           = 'AND (s.CenterID IS NULL OR s.CenterID=:sid) ';
```
done https://github.com/aces/Loris/issues/4267

ridz1208 commented:
@driusan I approved changes for the subset I'm responsible of only

driusan commented on modules/statistics/php/stats_mri.class.inc:
```diff
@@ -256,7 +256,7 @@ class Stats_MRI extends \NDB_Form
         }
         //SITES
         if ($_REQUEST['MRIsite'] ?? '') {
```
not sure in the context of an if statement.. I'm assuming someone did it intending to suppress a warning about $_REQUEST['MRIsite'] not being defined.

xlecours commented on modules/issue_tracker/ajax/EditIssue.php:
```diff
@@ -195,7 +195,7 @@ function validateInput($values)
             "
             SELECT
-                CenterID = :center_id
+                CenterID = :center_id as CenterID
```
Please change CenterID -> RegistrationCenterID

cmadjar replied:
@xlecours Good catch!! This is fixed now.

cmadjar commented:
@ridz1208 Sorry, the latest changes dismissed your review. Can you please review it again? Thanks!

ridz1208 commented:
@nicolasbrossard please review your section

xlecours commented:
@cmadjar The `testNotResolvedSearchButton` test in mri_violations module is failing Travis.
Other than that, everything seems fine.

nicolasbrossard commented:
The MRI violations module loads and looks OK, except for a display bug in the MRI violations results table that existed prior to Cecile's PR. Approving.

kongtiaowang commented on modules/mri_violations/php/mri_violations.class.inc:
```diff
@@ -246,7 +246,7 @@ class Mri_Violations extends \NDB_Menu_Filter_Form
             )
             LEFT JOIN psc p
-            ON (p.CenterID = c.CenterID)
+            ON (p.CenterID = s.CenterID)
```
It should be " ON (p.CenterID = c. RegistrationCenterID)"

kongtiaowang commented on modules/mri_violations/php/mri_violations.class.inc:
```diff
@@ -278,7 +278,7 @@ class Mri_Violations extends \NDB_Menu_Filter_Form
             ON (mrl.Visit_label = s.Visit_label AND mrl.CandID=s.CandID)
             LEFT JOIN psc p
-            ON (p.CenterID = c.CenterID)
+            ON (p.CenterID = s.CenterID)
```
It should be " ON (p.CenterID = c. RegistrationCenterID)"

kongtiaowang commented on modules/mri_violations/php/resolved_violations.class.inc:
```diff
@@ -233,7 +233,7 @@ class Resolved_Violations extends \NDB_Menu_Filter_Form
             ON (mri_protocol_violated_scans.CandID = s.CandID)
             LEFT JOIN psc p
-            ON (p.CenterID = c.CenterID)
+            ON (p.CenterID = s.CenterID)
```
It should be " ON (p.CenterID = c. RegistrationCenterID)"

kongtiaowang commented on modules/mri_violations/php/resolved_violations.class.inc:
```diff
@@ -260,7 +260,7 @@ class Resolved_Violations extends \NDB_Menu_Filter_Form
             ON (mri_violations_log.CandID = s.CandID)
             LEFT JOIN psc p
-            ON (p.CenterID = c.CenterID)
+            ON (p.CenterID = s.CenterID)
```
It should be " ON (p.CenterID = c. RegistrationCenterID)"

kongtiaowang commented on modules/mri_violations/php/resolved_violations.class.inc:
```diff
@@ -285,7 +285,7 @@ class Resolved_Violations extends \NDB_Menu_Filter_Form
             ON (c.CandID = s.CandID)
             LEFT JOIN psc p
-            ON (p.CenterID = c.CenterID)
+            ON (p.CenterID = s.CenterID)
```
It should be " ON (p.CenterID = c. RegistrationCenterID)"

kongtiaowang commented on modules/statistics/php/stats_mri.class.inc:
```diff
@@ -379,7 +379,7 @@ class Stats_MRI extends \NDB_Form
               GROUP BY Subcat,
                     s.SubprojectID,
-                    c.CenterID,
+                    s.CenterID,
```
c. RegistrationCenterID

ridz1208 commented:
@kongtiaowang I agree with @cmadjar s.centerid makes more sens here

ridz1208 commented:
@cmadjar if you manually test MRI violations is everything working fine ?

thats where travis seems to be having issues

cmadjar commented:
@ridz1208 Indeed, the MRI violation module seems to work OK aside from what @nicolasbrossard reported which is a bug on major unrelated to my changes:

> Fyi, from @nicolasbrossard :  The MRI violations module loads and looks OK, except for a display bug in the MRI violations results table that existed prior to Cecile's PR. Approving.

kongtiaowang commented:
@cmadjar , I made a change on your repo, you need to merge it. Sorry for the force update.
https://github.com/cmadjar/Loris/pull/3

ridz1208 commented:
@cmadjar 
`Parse error: syntax error, unexpected ';', expecting ')' in /app/modules/mri_violations/test/mri_violationsTest.php on line 90`

cmadjar commented:
@PapillonMcGill @zaliqarosli @nicolasbrossard @ridz1208 @xlecours A force push deleted your approval review. Could you look at that PR one last time? Thanks!

zaliqarosli commented:
@cmadjar i'm getting a lot of issues with the issue tracker :( so i'm going to properly test again

cmadjar commented:
@driusan yes pretty please :). This way I can start working on phase 2 this week!
